ledger/ledgercore/persistedacctdata.go

```diff
@@ -116,3 +117,458 @@ var EncodedMaxAppLocalStates = basics.EncodedMaxAppLocalStates
+	gd.Amounts[ai] = hodl.Amount
+	gd.Frozens[ai] = hodl.Frozen
+}
+
+func (gd *AssetsHoldingGroupData) delete(ai int) {
+	copy(gd.Amounts[ai:], gd.Amounts[ai+1:])
+	gd.Amounts = gd.Amounts[:len(gd.Amounts)-1]
+
+	copy(gd.Frozens[ai:], gd.Frozens[ai+1:])
+	gd.Frozens = gd.Frozens[:len(gd.Frozens)-1]
+
+	if ai == 0 {
+		gd.AssetOffsets = gd.AssetOffsets[1:]
+		gd.AssetOffsets[0] = 0
+	} else if ai == len(gd.AssetOffsets)-1 {
+		gd.AssetOffsets = gd.AssetOffsets[:len(gd.AssetOffsets)-1]
+	} else {
+		gd.AssetOffsets[ai+1] += gd.AssetOffsets[ai]
+		copy(gd.AssetOffsets[ai:], gd.AssetOffsets[ai+1:])
+		gd.AssetOffsets = gd.AssetOffsets[:len(gd.AssetOffsets)-1]
+	}
+}
+
+// GetHolding returns AssetHolding from group data by asset index ai
+func (gd AssetsHoldingGroupData) GetHolding(ai int) basics.AssetHolding {
+	return basics.AssetHolding{Amount: gd.Amounts[ai], Frozen: gd.Frozens[ai]}
+}
+
+// GetHolding returns AssetHolding from group data by asset index ai
+func (g AssetsHoldingGroup) GetHolding(ai int) basics.AssetHolding {
+	return g.groupData.GetHolding(ai)
+}
+
+// Encode returns msgp-encoded group data
+func (g AssetsHoldingGroup) Encode() []byte {
+	return protocol.Encode(&g.groupData)
+}
+
+// GetGroupData returns group data. Used in tests only
+func (g AssetsHoldingGroup) GetGroupData() AssetsHoldingGroupData {
+	return g.groupData
+}
+
+// Update an asset holding by index
+func (g *AssetsHoldingGroup) Update(ai int, holdings basics.AssetHolding) {
+	g.groupData.update(ai, holdings)
+}
+
+// Loaded return a boolean flag indicated if the group loaded or not
+func (g AssetsHoldingGroup) Loaded() bool {
+	return g.loaded
+}
+
+// Load sets a group data value in the group
+func (g *AssetsHoldingGroup) Load(gd AssetsHoldingGroupData) {
+	g.groupData = gd
+	g.loaded = true
+}
+
+func (g *AssetsHoldingGroup) delete(ai int) {
+	// although a group with only one element is handled by a caller
+	// add a safety check here
+	if g.Count == 1 {
+		*g = AssetsHoldingGroup{}
+		return
+	}
+
+	if ai == 0 {
+		// when deleting the first element, update MinAssetIndex
+		g.MinAssetIndex += g.groupData.AssetOffsets[1]
+	} else if uint32(ai) == g.Count-1 {
+		// when deleting the last element, update DeltaMaxAssetIndex
+		g.DeltaMaxAssetIndex -= uint64(g.groupData.AssetOffsets[len(g.groupData.AssetOffsets)-1])
+	}
+	g.groupData.delete(ai)
+	g.Count--
+}
+
+func (g *AssetsHoldingGroup) insert(aidx basics.AssetIndex, holding basics.AssetHolding) {
+	if aidx < g.MinAssetIndex {
+		// prepend
+		g.groupData.Amounts = append([]uint64{holding.Amount}, g.groupData.Amounts...)
+		g.groupData.Frozens = append([]bool{holding.Frozen}, g.groupData.Frozens...)
+		g.groupData.AssetOffsets[0] = g.MinAssetIndex - aidx
+		g.groupData.AssetOffsets = append([]basics.AssetIndex{0}, g.groupData.AssetOffsets...)
+		g.MinAssetIndex = aidx
+	} else if aidx >= g.MinAssetIndex+basics.AssetIndex(g.DeltaMaxAssetIndex) {
```

Comment:
to avoid handling existing entries both here and on the `else` case, make this into a `>`.